src/main/java/pl/javastart/couponcalc/PriceCalculator.java

```diff
@@ -1,13 +1,55 @@
 package pl.javastart.couponcalc;
 
-import java.util.List;
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.*;
+import java.util.stream.Collectors;
 
 public class PriceCalculator {
 
     public double calculatePrice(List<Product> products, List<Coupon> coupons) {
-        // TODO
+        if (Objects.isNull(products) && Objects.isNull(coupons)) {
+            return 0;
+        }
 
-        return 0;
+        if (Objects.isNull(coupons)) {
+            return products.stream()
+                    .mapToDouble(Product::getPrice)
+                    .sum();
+        }
+
+        Map<Category, Integer> categoryWithDiscountMap = coupons.stream()
+                .collect(Collectors.toMap(Coupon::getCategory, Coupon::getDiscountValueInPercents));
+
+        Map<Category, Double> categoryPriceMap = new HashMap<>();
+        for (Product product : products) {
+            categoryPriceMap.put(product.getCategory(), categoryPriceMap.getOrDefault(product.getCategory(), 0.0) + product.getPrice());
```

Comment:
Tutaj dobrze pasuje metoda merge (scalenie)
```
categoryPriceMap.merge(product.getCategory(), product.getPrice(), Double::sum);
```
Co oznacza, jeżeli kategoria nie istnieje, to wstaw do mapy kategorię i przypisz do niej cenę produktu, a jeżeli taka kategoria już istnieje, to dodaj cenę do istniejącej wartości